sematic/plugins/building/docker_builder.py

```diff
@@ -0,0 +1,496 @@
+        return build_config.base_uri
+
+    if build_config.image_script is None:
+        # appease mypy
+        raise BuildConfigurationError(
+            "Exactly one of `image_script` and `base_uri` must be specified!"
+        )
+
+    try:
+        script_dir, script_file = os.path.split(build_config.image_script)
+        if script_dir == "":
+            script_dir = None  # type: ignore
+        script_file = f"./{script_file}"
+
+        logger.info(
+            f"Executing: executable={script_file} cwd={script_dir} args={target}"
+        )
+
+        # the subprocess' stderr will be inherited from the current process,
+        # so it will print directly to the logs
+        with subprocess.Popen(
+            executable=script_file,
+            cwd=script_dir,
+            args=target,
+            stdout=subprocess.PIPE,
+            text=True,
+        ) as subproc:
+
+            image_uri, _ = subproc.communicate()
+
+            if subproc.returncode != 0:
+                raise SystemExit(subproc.returncode)
+
+            image_uri = image_uri.strip()
+
+    except BaseException as e:
+        raise BuildError(
+            f"Unable to source container image URI from '{build_config.image_script}'!"
+        ) from e
+
+    return ImageURI.from_uri(uri=image_uri)
+
+
+def _push_image(
+    image: docker.models.images.Image,  # type: ignore
+    image_uri: ImageURI,
+    push_config: Optional[ImagePushConfig],
+    docker_client: docker.DockerClient,  # type: ignore
+) -> ImageURI:
+    """
+    Re-tags and pushes the image to the configured repository.
+
+    Updates the `image` through lateral effect!
+
+    Parameters
+    ----------
+    image: docker.models.images.Image
+        The image to push
+    image_uri: ImageURI
+        The initial URI of the image to push
+    push_config: Optional[ImagePushConfig]
+        Instructions where to push the image
+    docker_client: docker.DockerClient
+        The `DockerClient` to use to execute the operations
+
+    Returns
+    -------
+    ImageURI:
+        The new `ImageURI` that indicates the location where the `image` was pushed.
+
+    Raises
+    ------
+    BuildError:
+        There was an error when executing the `docker` commands.
+    """
+    if push_config is None:
+        logger.info("Image pushing not configured; skipping")
+        return image_uri
+
+    repository = push_config.get_repository_str()
+    tag = push_config.get_tag()
+
+    tagging_successful = image.tag(repository=repository, tag=tag)
+
+    if not tagging_successful:
+        raise BuildError(
+            f"Tagging image '{image_uri}' in repository '{repository}' "
+            f"with tag '{tag}' failed; no other information available."
+        )
+
+    logger.info(
+        "Tagged image '%s' in repository '%s' with tag '%s'",
+        image_uri,
+        repository,
+        tag,
+    )
+
+    response = docker_client.images.push(
```

Comment:
How is the docker credential handled, please?

Reply:
The `docker_client = docker.from_env()` line of code uses the authentication details in `~/.docker/config.json`.
The documentation will mention that the local configuration is used.